### PR TITLE
WT-6284 History store isn't visible to wt dump

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
When we moved `WT_CURSTD_IGNORE_TOMBSTONE` from being a session flag to a cursor flag, we broke wt dump. `wt dump` uses a dump cursor which delegates calls to the cursor API to a "child" cursor. We need to set the flag on the child cursor for this to work.

We also modified the check in `__wt_txn_read` such that we will read the on-disk value's stop time pair even with `WT_CURSTD_IGNORE_TOMBSTONE` enabled if it is globally visible. This makes `wt dump` less useful so let's make an exception here.

### CODE REVIEW

File: src/utilities/util_dump.c [hunk 1]
```diff
@@ -154,8 +155,11 @@ util_dump(WT_SESSION *session, int argc, char *argv[])
          * nothing will be visible. The only exception is if we've supplied a timestamp in which
          * case, we're specifically interested in what is visible at a given read timestamp.
          */
-        if (WT_STREQ(simpleuri, WT_HS_URI) && timestamp == NULL)
-            F_SET(cursor, WT_CURSTD_IGNORE_TOMBSTONE);
+        if (WT_STREQ(simpleuri, WT_HS_URI) && timestamp == NULL) {
+            dump_cursor = (WT_CURSOR_DUMP *)cursor;
+            /* Set the "ignore tombstone" flag on the underlying cursor. */
+            F_SET(dump_cursor->child, WT_CURSTD_IGNORE_TOMBSTONE);
+        }
```
Comment (quchenhao) [inline src/utilities/util_dump.c:162]: Should we change this line as well?

Comment (tetsuo-cpp) [reply src/utilities/util_dump.c:162]: Hmm, I'm not sure why we bother doing that. I may as well though.

**Comment by tetsuo-cpp** [comment]
> What's the point of dumping entries in the history store that are already obsolete?

In my case, I was debugging history store for a workload that wasn't using timestamps. I think what happened was that because the `txnid` of the tombstones got zeroed out and there is no timestamp, I wasn't able to see anything even though the history store was relatively large.

Also, generally speaking, I use this for debugging issues. So even if the record is obsolete, the key still exists internally and affects WiredTiger (in my case, I was trying to diagnose out of order keys).

File: src/utilities/util_dump.c (hunk 1 above)
Comment (tetsuo-cpp) [reply src/utilities/util_dump.c:163]: f32d3d5

**Comment by quchenhao** [comment]
> > What's the point of dumping entries in the history store that are already obsolete?
> 
> In my case, I was debugging history store for a workload that wasn't using timestamps. I think what happened was that because the `txnid` of the tombstones got zeroed out and there is no timestamp, I wasn't able to see anything even though the history store was relatively large.
> 
> Also, generally speaking, I use this for debugging issues. So even if the record is obsolete, the key still exists internally and affects WiredTiger (in my case, I was trying to diagnose out of order keys).

Can we make that an option? I think dumping obsolete data in most of the time just causes confusion.

**Comment by tetsuo-cpp** [comment]
> > > What's the point of dumping entries in the history store that are already obsolete?
> > 
> > 
> > In my case, I was debugging history store for a workload that wasn't using timestamps. I think what happened was that because the `txnid` of the tombstones got zeroed out and there is no timestamp, I wasn't able to see anything even though the history store was relatively large.
> > Also, generally speaking, I use this for debugging issues. So even if the record is obsolete, the key still exists internally and affects WiredTiger (in my case, I was trying to diagnose out of order keys).
> 
> Can we make that an option? I think dumping obsolete data in most of the time just causes confusion.

It doesn't work for non-timestamped tables so it can't be the default option. I don't agree that it's confusing compared to what you're suggesting. This way, the stop time pair is treated consistently by `dump` whereas currently, it is ignored but only sometimes.

**Comment by quchenhao** [comment]
Why it doesn't work for non-timestamped tables? I think it is confusing is because you don't know which record should been seen and which not. You need to figure it out by yourself.

**Comment by tetsuo-cpp** [comment]
> Why it doesn't work for non-timestamped tables?

The transaction id is getting wiped so I think for non-timestamped tables, the stop pairs are always going to be globally visible. That makes sense for recovery but it makes `dump` less useful.

> I think it is confusing is because you don't know which record should been seen and which not. You need to figure it out by yourself.

Yeah but that's true for any history store record printed by `dump`. A lot of them have stop pairs so it's not obvious what would be visible for a given reader. That's why we have the `-t` flag for dumping at a given timestamp.

**Comment by quchenhao** [comment]
How we wiped the txnid? We should only do that after reopen, right?

**Comment by quchenhao** [comment]
I get it. dump is after reopen

**Pull request merged**